src/functor.jl

```diff
@@ -169,7 +169,54 @@ function _default_walk(f, x)
   func, re = functor(x)
   re(map(f, func))
 end
-_default_walk(f, ::Nothing, ::Nothing) = nothing
+_default_walk(_, ::Nothing, ::Nothing) = nothing
+
+# Side effects only, saves a restructure
+function _foreach_walk(f, x)
+  foreach(f, children(x))
+  return x
+end
+
+### WARNING: the following is unstable internal functionality. Use at your own risk!
+# Wrapper over an IdDict which only saves values with a stable object identity
+struct Cache{K,V}
+  inner::IdDict{K,V}
+end
+Cache() = Cache(IdDict())
+
+iscachesafe(x) = !isbits(x) && ismutable(x)
```

Comment:
Can I suggest calling this `usecache`? It's not so clear to me which way "safe" points, safe that we don't need the cache, or safe that we can use it? 

Also, why `ismutable`? This catches only the outermost wrapper, which I don't think ought to matter.

Reply:
`usecache` sounds better, changed.

`ismutable` is required because this function is called at every level of the tree. So if there is a string in a (possibly deeply) nested struct that otherwise contains all bitstype values, `isbits` will fail and we will get a false positive. In fact `ismutable` may be sufficient on its own, but in the spirit of not making too many assumptions around underspecified language behaviour (namely that all mutable structs have stable objectids) I've kept it in to be safe.